apps/system/js/advanced_telemetry.js

```diff
@@ -0,0 +1,426 @@
+      }
+    }, self.metrics.getInterval());
+  };
+
+  AT.prototype.stopBatch = function stopBatch() {
+    clearTimeout(this.timer);
+  };
+
+  // Once we have the histogram payload, we can send it to the telemetry
+  // server.  Once we receive 200 OK back from telemetry server, we can
+  // start a fresh batch of metrics.
+  AT.prototype.transmit = function transmit(payload) {
+    var self = this;
+
+    if (!this.collecting) {
+      return;
+    }
+
+    // But assume that it will succeed and start collecting new metrics now
+    this.metrics = new BatchTiming(false);
+    this.startBatch();
+
+    var deviceInfoQuery = {
+      'deviceinfo.hardware': 'unknown',
+      'app.update.channel': 'unknown',
+      'deviceinfo.platform_version': 'unknown',
+      'deviceinfo.platform_build_id': 'unknown'
+    };
+
+
+    // Query the settings db for parameters for the URL
+    AT.getSettings(deviceInfoQuery, function(deviceResponse) {
+      // Note that this wrapper is using the new v4 Unified Telemetry format
+      var wrapper = {
+        type: REASON,
+        id: uuid(),
+        creationDate: new Date().toISOString(),
+        version: AT.TELEMETRY_VERSION,
+        application: {
+          architecture: 'x86',
+          buildId: deviceResponse['deviceinfo.platform_build_id'],
+          name: AT.TELEMETRY_APP_NAME,
+          version: deviceResponse['deviceinfo.platform_version'],
+          vendor: 'Mozilla',
+          platformVersion: deviceResponse['deviceinfo.platform_version'],
+          xpcomAbi: 'x86-msvc',
+          channel: deviceResponse['app.update.channel']
+        },
+        clientId: self.deviceID,
+        payload: payload
+      };
+
+      debuglongLine(JSON.stringify(wrapper));
+
+      // Build the wrapper for the telemetry version
+      send(wrapper, deviceResponse);
+    });
+
+    function send(payload, deviceInfoQuery) {
+      var request = new AdvancedTelemetryPing(payload, deviceInfoQuery,
+        self.deviceID);
+
+      // We don't actually have to do anything if the data is transmitted
+      // successfully. We are already set up to collect the next batch of data.
+      function onload() {
+        loginfo('Transmitted Successfully.');
+        // TODO: Add this line below to AdvancedTelemetryHelper as an API.
+        console.info('telemetry|MGMT|CLEARMETRICS');
+      }
+
+      function retry(e) {
+        // If the attempt to transmit a batch of data fails, refresh the payload
+        loginfo('App usage metrics transmission failure:', e.type);
+        self.getPayload();
+      }
+
+      request.send({
+        timeout: AT.REPORT_TIMEOUT,
+        onload: onload,
+        onerror: retry,
+        onabort: retry,
+        ontimeout: retry
+      });
+    }
+  };
+
+  function AdvancedTelemetryPing(payload, deviceQuery, did) {
+    if (!payload) {
+      throw new Error('No arguments');
+    }
+    // clone so we don't put data into the object that was given to us
+    this.packet = payload;
```

Comment:
This isn't actually a clone, just a reference. If you want to copy the data, the most straight forward way would be:

`this.packet = JSON.parse(JSON.stringify(payload));`